internal/provider/resource_gitlab_project_environment.go

```diff
@@ -0,0 +1,182 @@
+
+	client := meta.(*gitlab.Client)
+
+	environment, _, err := client.Environments.CreateEnvironment(project, &options, gitlab.WithContext(ctx))
+	if err != nil {
+		if is404(err) {
+			return diag.Errorf("feature Environments is not available")
+		}
+		return diag.FromErr(err)
+	}
+
+	environmentID := fmt.Sprintf("%d", environment.ID)
+	d.SetId(buildTwoPartID(&project, &environmentID))
+
+	return resourceGitlabProjectEnvironmentRead(ctx, d, meta)
+}
+
+func resourceGitlabProjectEnvironmentRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	log.Printf("[DEBUG] read gitlab environment %s", d.Id())
+
+	project, environmentID, environmentIDInt, err := environmentAndProjectFromResource(d)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	log.Printf("[DEBUG] Project %s read gitlab environment %d", project, environmentIDInt)
+
+	client := meta.(*gitlab.Client)
+
+	environment, _, err := client.Environments.GetEnvironment(project, environmentIDInt, gitlab.WithContext(ctx))
+	if err != nil {
+		if is404(err) {
+			log.Printf("[DEBUG] Project %s gitlab environment %q not found, removing from state", project, environmentID)
+			d.SetId("")
+			return nil
+		}
+		return diag.Errorf("error getting gitlab project %q environment %q: %v", project, environmentID, err)
+	}
+
+	d.Set("project", project)
+	d.Set("name", environment.Name)
+	d.Set("state", environment.State)
+	d.Set("external_url", environment.ExternalURL)
+
+	return nil
+}
+
+func resourceGitlabProjectEnvironmentUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	log.Printf("[DEBUG] update gitlab environment %s", d.Id())
+
+	project, environmentID, environmentIDInt, err := environmentAndProjectFromResource(d)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	options := &gitlab.EditEnvironmentOptions{
+		Name: gitlab.String(d.Get("name").(string)),
+	}
+
+	if d.HasChange("external_url") {
+		options.ExternalURL = gitlab.String(d.Get("external_url").(string))
+	}
+
+	log.Printf("[DEBUG] Project %s update gitlab environment %d", project, environmentIDInt)
+
+	client := meta.(*gitlab.Client)
+
+	if _, _, err := client.Environments.EditEnvironment(project, environmentIDInt, options, gitlab.WithContext(ctx)); err != nil {
+		return diag.Errorf("error editing gitlab project %q environment %q: %v", project, environmentID, err)
+	}
+
+	return resourceGitlabProjectEnvironmentRead(ctx, d, meta)
+}
+
+func resourceGitlabProjectEnvironmentDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	project, environmentID, environmentIDInt, err := environmentAndProjectFromResource(d)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	log.Printf("[DEBUG] Project %s delete gitlab project-level environment %v", project, environmentIDInt)
+
+	client := meta.(*gitlab.Client)
+
+	_, err = client.Environments.StopEnvironment(project, environmentIDInt, gitlab.WithContext(ctx))
```

Comment:
Oh I did not even notice that the previous implementer tried to stop the environment before a DELETE happened.  I have just experimented and it appears that if you try to delete an environment without it being in a stopped state you get a 403, a deployment does not need to happen ever in order to put an environment in a stopped state. With this information I propose if we do a flag at all we document what happens when you delete an environment or recommend the use of the [`prevent_destroy`](https://www.terraform.io/language/meta-arguments/lifecycle#prevent_destroy) meta argument to prevent an environment from being stopped & deleted or we do more of an S3 bucket [`force_destroy`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/s3_bucket#force_destroy) attribute

Reply:
hmmm, good question. I think because the stopping is outside of the normal expected CRUD functions a resource performs I wouldn't favor just recommending the `prevent_destroy` meta argument ... wdyt? 

The `force_destroy` attribute basically is the `stop_before_destroy` w.r.t behavior right? Just the name is different?
If so, regarding naming I'd prefer `stop_before_destroy` because it is more informative about what is enforced there, in contrary `force_destroy` does not give any information about what is being done to enforce the destroy. Does that make sense? 😅

Reply:
Implemented